{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez, GenBank, SeqIO\n",
    "Entrez.email = \"gsyu93@gmail.com\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "handle = Entrez.esearch(db=\"clinvar\", term='VCV000209223', idtype='acc')\n",
    "result = Entrez.read(handle)\n",
    "\n",
    "handle.close()\n",
    "# records = Entrez.parse(handle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Count': '1', 'RetMax': '1', 'RetStart': '0', 'IdList': ['209223'], 'TranslationSet': [], 'TranslationStack': [{'Term': 'VCV000209223[All Fields]', 'Field': 'All Fields', 'Count': '1', 'Explode': 'N'}, 'GROUP'], 'QueryTranslation': 'VCV000209223[All Fields]'}"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "b'<?xml version=\"1.0\" encoding=\"UTF-8\" ?>\\n                        \\n<ClinVarResult-Set><VariationArchive VariationID=\"209223\" VariationName=\"NM_007294.3(BRCA1):c.*4549G&gt;C\" VariationType=\"single nucleotide variant\" Accession=\"VCV000209223\" Version=\"3\" RecordType=\"interpreted\" NumberOfSubmissions=\"1\" NumberOfSubmitters=\"1\" DateLastUpdated=\"2022-11-05\" DateCreated=\"2015-09-29\" MostRecentSubmission=\"2015-09-29\">\\n  <RecordStatus>current</RecordStatus>\\n  <Species>Homo sapiens</Species>\\n  <InterpretedRecord>\\n    <SimpleAllele AlleleID=\"206181\" VariationID=\"209223\">\\n      <GeneList>\\n        <Gene Symbol=\"BRCA1\" FullName=\"BRCA1 DNA repair associated\" GeneID=\"672\" HGNC_ID=\"HGNC:1100\" Source=\"submitted\" RelationshipType=\"near gene, downstream\">\\n          <Location>\\n            <CytogeneticLocation>17q21.31</CytogeneticLocation>\\n            <SequenceLocation Assembly=\"GRCh38\" AssemblyAccessionVersion=\"GCF_000001405.38\" AssemblyStatus=\"current\" Chr=\"17\" Accession=\"NC_000017.11\" start=\"43044295\" stop=\"43125364\" display_start=\"43044295\" display_stop=\"43125364\" Strand=\"-\"/>\\n            <SequenceLocation Assembly=\"GRCh37\" AssemblyAccessionVersion=\"GCF_000001405.25\" AssemblyStatus=\"previous\" Chr=\"17\" Accession=\"NC_000017.10\" start=\"41196311\" stop=\"41277499\" display_start=\"41196311\" display_stop=\"41277499\" Strand=\"-\"/>\\n          </Location>\\n          <OMIM>113705</OMIM>\\n          <Haploinsufficiency last_evaluated=\"2021-09-23\" ClinGen=\"https://www.ncbi.nlm.nih.gov/projects/dbvar/ISCA/isca_gene.cgi?sym=BRCA1\">Sufficient evidence for dosage pathogenicity</Haploinsufficiency>\\n          <Triplosensitivity last_evaluated=\"2021-09-23\" ClinGen=\"https://www.ncbi.nlm.nih.gov/projects/dbvar/ISCA/isca_gene.cgi?sym=BRCA1\">No evidence available</Triplosensitivity>\\n          <Property>gene_acmg_incidental_2013</Property>\\n          <Property>gene_acmg_incidental_2016</Property>\\n          <Property>gene_acmg_incidental_2021</Property>\\n          <Property>gene_acmg_incidental_2022</Property>\\n        </Gene>\\n      </GeneList>\\n      <Name>NM_007294.3(BRCA1):c.*4549G&gt;C</Name>\\n      <CanonicalSPDI>NC_000017.11:43041128:C:G</CanonicalSPDI>\\n      <VariantType>single nucleotide variant</VariantType>\\n      <Location>\\n        <CytogeneticLocation>17q21.31</CytogeneticLocation>\\n        <SequenceLocation Assembly=\"GRCh38\" AssemblyAccessionVersion=\"GCF_000001405.38\" forDisplay=\"true\" AssemblyStatus=\"current\" Chr=\"17\" Accession=\"NC_000017.11\" start=\"43041129\" stop=\"43041129\" display_start=\"43041129\" display_stop=\"43041129\" variantLength=\"1\" positionVCF=\"43041129\" referenceAlleleVCF=\"C\" alternateAlleleVCF=\"G\"/>\\n        <SequenceLocation Assembly=\"GRCh37\" AssemblyAccessionVersion=\"GCF_000001405.25\" AssemblyStatus=\"previous\" Chr=\"17\" Accession=\"NC_000017.10\" start=\"41193146\" stop=\"41193146\" display_start=\"41193146\" display_stop=\"41193146\" variantLength=\"1\" positionVCF=\"41193146\" referenceAlleleVCF=\"C\" alternateAlleleVCF=\"G\"/>\\n      </Location>\\n      <OtherNameList>\\n        <Name>10260 G&gt;C</Name>\\n      </OtherNameList>\\n      <HGVSlist>\\n        <HGVS Type=\"genomic\">\\n          <NucleotideExpression sequenceAccessionVersion=\"LRG_292\" sequenceAccession=\"LRG_292\">\\n            <Expression>LRG_292:g.176855G&gt;C</Expression>\\n          </NucleotideExpression>\\n        </HGVS>\\n        <HGVS Type=\"coding\">\\n          <NucleotideExpression sequenceAccessionVersion=\"LRG_292t1\" sequenceAccession=\"LRG_292t1\">\\n            <Expression>LRG_292t1:c.*4549G&gt;C</Expression>\\n          </NucleotideExpression>\\n        </HGVS>\\n        <HGVS Assembly=\"GRCh37\" Type=\"genomic, top-level\">\\n          <NucleotideExpression sequenceAccessionVersion=\"NC_000017.10\" sequenceAccession=\"NC_000017\" sequenceVersion=\"10\" change=\"g.41193146C&gt;G\" Assembly=\"GRCh37\">\\n            <Expression>NC_000017.10:g.41193146C&gt;G</Expression>\\n          </NucleotideExpression>\\n        </HGVS>\\n        <HGVS Assembly=\"GRCh38\" Type=\"genomic, top-level\">\\n          <NucleotideExpression sequenceAccessionVersion=\"NC_000017.11\" sequenceAccession=\"NC_000017\" sequenceVersion=\"11\" change=\"g.43041129C&gt;G\" Assembly=\"GRCh38\">\\n            <Expression>NC_000017.11:g.43041129C&gt;G</Expression>\\n          </NucleotideExpression>\\n        </HGVS>\\n        <HGVS Type=\"genomic\">\\n          <NucleotideExpression sequenceAccessionVersion=\"NG_005905.2\" sequenceAccession=\"NG_005905\" sequenceVersion=\"2\" change=\"g.176855G&gt;C\">\\n            <Expression>NG_005905.2:g.176855G&gt;C</Expression>\\n          </NucleotideExpression>\\n        </HGVS>\\n      </HGVSlist>\\n      <XRefList>\\n        <XRef ID=\"CA276199\" DB=\"ClinGen\"/>\\n        <XRef Type=\"rs\" ID=\"138528005\" DB=\"dbSNP\"/>\\n      </XRefList>\\n      <AlleleFrequencyList>\\n        <AlleleFrequency Value=\"0.00158\" Source=\"The Genome Aggregation Database (gnomAD)\"/>\\n        <AlleleFrequency Value=\"0.00149\" Source=\"Trans-Omics for Precision Medicine (TOPMed)\"/>\\n        <AlleleFrequency Value=\"0.00200\" Source=\"1000 Genomes Project\"/>\\n      </AlleleFrequencyList>\\n      <GlobalMinorAlleleFrequency Value=\"0.00200\" Source=\"1000 Genomes Project\" MinorAllele=\"G\"/>\\n    </SimpleAllele>\\n    <ReviewStatus>reviewed by expert panel</ReviewStatus>\\n    <RCVList>\\n      <RCVAccession Title=\"NM_007294.3(BRCA1):c.*4549G&gt;C AND Breast-ovarian cancer, familial, susceptibility to, 1\" DateLastEvaluated=\"2015-01-12\" ReviewStatus=\"reviewed by expert panel\" Interpretation=\"Benign\" SubmissionCount=\"1\" Accession=\"RCV000191166\" Version=\"3\">\\n        <InterpretedConditionList TraitSetID=\"1917\">\\n          <InterpretedCondition DB=\"MedGen\" ID=\"C2676676\">Breast-ovarian cancer, familial, susceptibility to, 1</InterpretedCondition>\\n        </InterpretedConditionList>\\n      </RCVAccession>\\n    </RCVList>\\n    <Interpretations>\\n      <Interpretation DateLastEvaluated=\"2015-01-12\" NumberOfSubmissions=\"1\" NumberOfSubmitters=\"1\" Type=\"Clinical significance\">\\n        <Description>Benign</Description>\\n        <ConditionList>\\n          <TraitSet ID=\"1917\" Type=\"Disease\" ContributesToAggregateClinsig=\"true\">\\n            <Trait ID=\"4711\" Type=\"Disease\">\\n              <Name>\\n                <ElementValue Type=\"Alternate\">Breast-ovarian cancer, familial 1</ElementValue>\\n                <XRef ID=\"Breast-ovarian+cancer%2C+familial+1/7865\" DB=\"Genetic Alliance\"/>\\n              </Name>\\n              <Name>\\n                <ElementValue Type=\"Alternate\">BREAST CANCER, FAMILIAL, SUSCEPTIBILITY TO, 1</ElementValue>\\n                <XRef ID=\"604370\" DB=\"OMIM\"/>\\n              </Name>\\n              <Name>\\n                <ElementValue Type=\"Alternate\">Breast cancer, familial 1</ElementValue>\\n              </Name>\\n              <Name>\\n                <ElementValue Type=\"Alternate\">OVARIAN CANCER, SUSCEPTIBILITY TO</ElementValue>\\n                <XRef Type=\"Allelic variant\" ID=\"602667.0001\" DB=\"OMIM\"/>\\n              </Name>\\n              <Name>\\n                <ElementValue Type=\"Preferred\">Breast-ovarian cancer, familial, susceptibility to, 1</ElementValue>\\n                <XRef ID=\"MONDO:0011450\" DB=\"MONDO\"/>\\n              </Name>\\n              <Symbol>\\n                <ElementValue Type=\"Alternate\">HBOC1</ElementValue>\\n                <XRef Type=\"MIM\" ID=\"604370\" DB=\"OMIM\"/>\\n              </Symbol>\\n              <Symbol>\\n                <ElementValue Type=\"Alternate\">BROVCA1</ElementValue>\\n                <XRef Type=\"MIM\" ID=\"604370\" DB=\"OMIM\"/>\\n              </Symbol>\\n              <Symbol>\\n                <ElementValue Type=\"Alternate\">HBOC</ElementValue>\\n                <XRef ID=\"GTR000501743\" DB=\"Genetic Testing Registry (GTR)\"/>\\n              </Symbol>\\n              <AttributeSet>\\n                <Attribute Type=\"public definition\">BRCA1- and BRCA2-associated hereditary breast and ovarian cancer (HBOC) is characterized by an increased risk for female and male breast cancer, ovarian cancer (including fallopian tube and primary peritoneal cancers), and to a lesser extent other cancers such as prostate cancer, pancreatic cancer, and melanoma primarily in individuals with a BRCA2 pathogenic variant. The risk of developing an associated cancer varies depending on whether HBOC is caused by a BRCA1 or BRCA2 pathogenic variant.</Attribute>\\n                <XRef ID=\"NBK1247\" DB=\"GeneReviews\"/>\\n              </AttributeSet>\\n              <AttributeSet>\\n                <Attribute Type=\"keyword\">Neoplasm</Attribute>\\n              </AttributeSet>\\n              <AttributeSet>\\n                <Attribute Type=\"disease mechanism\" integerValue=\"273\">loss of function</Attribute>\\n                <XRef ID=\"GTR000501746\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000507864\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000509450\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000509692\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000509982\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000512644\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000512645\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000520872\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000552245\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000569406\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000021517\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000320777\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000332529\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000509001\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000509983\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000512320\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000520865\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000520871\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000522160\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000527942\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000530120\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000552304\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000562228\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000569423\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000570011\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000325409\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000501817\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000505644\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000507653\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000512816\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000514601\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000519030\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000520069\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000520869\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000522161\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000528915\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000531275\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000551440\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000017876\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000330054\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000501743\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000507764\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000509002\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000509980\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000520071\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000520072\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000520394\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000520866\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000520867\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000520870\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000521908\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000522159\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000530202\" DB=\"Genetic Testing Registry (GTR)\"/>\\n                <XRef ID=\"GTR000592961\" DB=\"Genetic Testing Registry (GTR)\"/>\\n              </AttributeSet>\\n              <Citation Type=\"review\" Abbrev=\"GeneReviews\">\\n                <ID Source=\"PubMed\">20301425</ID>\\n                <ID Source=\"BookShelf\">NBK1247</ID>\\n              </Citation>\\n              <Citation Type=\"practice guideline\" Abbrev=\"ACS, 2007\">\\n                <ID Source=\"PubMed\">17392385</ID>\\n              </Citation>\\n              <Citation Type=\"Position Statement\" Abbrev=\"ASCO, 2010\">\\n                <ID Source=\"PubMed\">20065170</ID>\\n              </Citation>\\n              <Citation Type=\"practice guideline\" Abbrev=\"NSGC, 2004\">\\n                <ID Source=\"PubMed\">15604628</ID>\\n              </Citation>\\n              <Citation Type=\"practice guideline\" Abbrev=\"NSGC, 2007\">\\n                <ID Source=\"PubMed\">17508274</ID>\\n              </Citation>\\n              <Citation Type=\"practice guideline\" Abbrev=\"ACOG, 2009\">\\n                <ID Source=\"PubMed\">19305347</ID>\\n              </Citation>\\n              <Citation Type=\"Recommendation\" Abbrev=\"ACMG, 2013\">\\n                <ID Source=\"PubMed\">23788249</ID>\\n              </Citation>\\n              <Citation Type=\"Suggested Reading\" Abbrev=\"Phillips et al., 2013\">\\n                <ID Source=\"PubMed\">23918944</ID>\\n              </Citation>\\n              <Citation Type=\"Suggested Reading\" Abbrev=\"Domchek et al., 2010\">\\n                <ID Source=\"pmc\">2948529</ID>\\n              </Citation>\\n              <Citation Type=\"practice guideline\" Abbrev=\"ASCO, 2014\">\\n                <ID Source=\"PubMed\">24493721</ID>\\n              </Citation>\\n              <Citation Type=\"practice guideline\" Abbrev=\"USPSTF, 2014 [Archived]\">\\n                <ID Source=\"PubMed\">24366376</ID>\\n              </Citation>\\n              <Citation Type=\"practice guideline\" Abbrev=\"NSGC, 2013\">\\n                <ID Source=\"PubMed\">23188549</ID>\\n              </Citation>\\n              <Citation Type=\"practice guideline\" Abbrev=\"ACMG/NSGC, 2015\">\\n                <ID Source=\"PubMed\">25394175</ID>\\n              </Citation>\\n              <Citation Type=\"Recommendation\" Abbrev=\"USPSTF, 2014\">\\n                <ID Source=\"PubMed\">24366402</ID>\\n              </Citation>\\n              <Citation Type=\"Suggested Reading\" Abbrev=\"USPSTF, 2013\">\\n                <ID Source=\"PubMed\">24432435</ID>\\n              </Citation>\\n              <Citation Type=\"Recommendation\" Abbrev=\"ACMG, 2015\">\\n                <ID Source=\"PubMed\">25356965</ID>\\n              </Citation>\\n              <Citation Type=\"Recommendation\" Abbrev=\"ACMG, 2016\">\\n                <ID Source=\"PubMed\">27854360</ID>\\n                <ID Source=\"DOI\">10.1038/gim.2016.190</ID>\\n              </Citation>\\n              <Citation Type=\"Suggested Reading\" Abbrev=\"NCI PDQ, Cancer Genetics Counseling\">\\n                <ID Source=\"PubMed\">26389258</ID>\\n              </Citation>\\n              <Citation Type=\"Recommendation\" Abbrev=\"ACMG SF v3.0, 2021\">\\n                <ID Source=\"PubMed\">34012068</ID>\\n                <ID Source=\"DOI\">10.1038/s41436-021-01172-3</ID>\\n              </Citation>\\n              <Citation Type=\"practice guideline\" Abbrev=\"NSGC, 2021\">\\n                <ID Source=\"PubMed\">33410258</ID>\\n              </Citation>\\n              <Citation Type=\"Recommendation\" Abbrev=\"ACMG SF v3.1, 2022\">\\n                <ID Source=\"PubMed\">35802134</ID>\\n                <ID Source=\"DOI\">10.1016/j.gim.2022.04.006</ID>\\n              </Citation>\\n              <Citation Type=\"Position Statement\" Abbrev=\"ASCO, 2015\">\\n                <ID Source=\"PubMed\">26324357</ID>\\n              </Citation>\\n              <Citation Type=\"practice guideline\" Abbrev=\"NCCN, 2022\">\\n                <URL>https://www.nccn.org/professionals/physician_gls/pdf/genetics_bop.pdf</URL>\\n                <CitationText>NCCN Clinical Practice Guidelines in Oncology, Genetic/Familial High-Risk Assessment: Breast, Ovarian, and Pancreatic version 2.2022</CitationText>\\n              </Citation>\\n              <Citation Type=\"Recommendation\" Abbrev=\"USPSTF, 2019\">\\n                <ID Source=\"PubMed\">31429903</ID>\\n              </Citation>\\n              <Citation Type=\"Suggested Reading\" Abbrev=\"USPSTF, 2019\">\\n                <ID Source=\"PubMed\">31479213</ID>\\n                <ID Source=\"BookShelf\">NBK545867</ID>\\n              </Citation>\\n              <XRef ID=\"145\" DB=\"Orphanet\"/>\\n              <XRef ID=\"C2676676\" DB=\"MedGen\"/>\\n              <XRef ID=\"MONDO:0011450\" DB=\"MONDO\"/>\\n              <XRef Type=\"MIM\" ID=\"604370\" DB=\"OMIM\"/>\\n            </Trait>\\n          </TraitSet>\\n        </ConditionList>\\n      </Interpretation>\\n    </Interpretations>\\n    <ClinicalAssertionList>\\n      <ClinicalAssertion ID=\"456000\" SubmissionDate=\"2015-08-17\" DateLastUpdated=\"2015-09-29\" DateCreated=\"2015-09-29\">\\n        <ClinVarSubmissionID localKey=\"NM_007294.3:c.*4549G&gt;C|OMIM:604370\"/>\\n        <ClinVarAccession Accession=\"SCV000244504\" DateUpdated=\"2015-09-29\" DateCreated=\"2015-09-29\" Type=\"SCV\" Version=\"1\" SubmitterName=\"Evidence-based Network for the Interpretation of Germline Mutant Alleles (ENIGMA)\" OrgID=\"504863\" OrganizationCategory=\"consortium\" OrgAbbreviation=\"ENIGMA\"/>\\n        <RecordStatus>current</RecordStatus>\\n        <ReviewStatus>reviewed by expert panel</ReviewStatus>\\n        <Interpretation DateLastEvaluated=\"2015-01-12\">\\n          <Description>Benign</Description>\\n          <Comment>Class 1 not pathogenic based on frequency &gt;1% in an outbred sampleset. Frequency 0.02 (African), derived from 1000 genomes (2012-04-30).</Comment>\\n        </Interpretation>\\n        <Assertion>variation to disease</Assertion>\\n        <AttributeSet>\\n          <Attribute Type=\"AssertionMethod\">ENIGMA BRCA1/2 Classification Criteria (2015)</Attribute>\\n          <Citation Type=\"general\">\\n            <URL>https://submit.ncbi.nlm.nih.gov/ft/byid/c2ffjci6/enigma_rules_2015-03-26.pdf</URL>\\n            <CitationText>ENIGMA BRCA1/2 Classification Criteria (2015)</CitationText>\\n          </Citation>\\n        </AttributeSet>\\n        <ObservedInList>\\n          <ObservedIn>\\n            <Sample>\\n              <Origin>germline</Origin>\\n              <Species TaxonomyId=\"9606\">human</Species>\\n              <AffectedStatus>unknown</AffectedStatus>\\n            </Sample>\\n            <Method>\\n              <MethodType>curation</MethodType>\\n            </Method>\\n            <ObservedData>\\n              <Attribute Type=\"Description\">not provided</Attribute>\\n            </ObservedData>\\n          </ObservedIn>\\n        </ObservedInList>\\n        <SimpleAllele>\\n          <GeneList>\\n            <Gene Symbol=\"BRCA1\"/>\\n          </GeneList>\\n          <VariantType>Variation</VariantType>\\n          <OtherNameList>\\n            <Name>10260 G&gt;C</Name>\\n          </OtherNameList>\\n          <AttributeSet>\\n            <Attribute Type=\"HGVS\">NM_007294.3:c.*4549G&gt;C</Attribute>\\n          </AttributeSet>\\n        </SimpleAllele>\\n        <TraitSet Type=\"Disease\">\\n          <Trait Type=\"Disease\">\\n            <Name>\\n              <ElementValue Type=\"Preferred\">Breast-ovarian cancer, familial 1</ElementValue>\\n            </Name>\\n            <XRef DB=\"OMIM\" ID=\"604370\" Type=\"MIM\"/>\\n          </Trait>\\n        </TraitSet>\\n        <StudyDescription>ENIGMA (Evidence-based Network for the Interpretation of Germline Mutant Alleles) is a consortium focused on determining the clinical significance of variants in BRCA1, BRCA2 and other known/suspected breast cancer genes. http://enigmaconsortium.org/</StudyDescription>\\n        <SubmissionNameList>\\n          <SubmissionName>ENIGMA_2015-08-17</SubmissionName>\\n        </SubmissionNameList>\\n      </ClinicalAssertion>\\n    </ClinicalAssertionList>\\n    <TraitMappingList>\\n      <TraitMapping ClinicalAssertionID=\"456000\" TraitType=\"Disease\" MappingType=\"Name\" MappingValue=\"Breast-ovarian cancer, familial 1\" MappingRef=\"Preferred\">\\n        <MedGen CUI=\"C2676676\" Name=\"Breast-ovarian cancer, familial, susceptibility to, 1\"/>\\n      </TraitMapping>\\n    </TraitMappingList>\\n  </InterpretedRecord>\\n</VariationArchive>\\n</ClinVarResult-Set>'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "handle = Entrez.efetch(db='clinvar', id='209223', rettype='vcv', is_varationid='true', from_esearch=\"true\")\n",
    "handle.read()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "handle = Entrez.efetch(db='clinvar', id='VCV000209223', rettype='vcv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import xml.etree.ElementTree as ET\n",
    "result = ET.parse(handle)\n",
    "root = result.getroot()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<ClinVarResult-Set>\n",
      "\t<VariationArchive VariationID=\"209223\" VariationName=\"NM_007294.3(BRCA1):c.*4549G&gt;C\" VariationType=\"single nucleotide variant\" Accession=\"VCV000209223\" Version=\"3\" RecordType=\"interpreted\" NumberOfSubmissions=\"1\" NumberOfSubmitters=\"1\" DateLastUpdated=\"2022-11-05\" DateCreated=\"2015-09-29\" MostRecentSubmission=\"2015-09-29\">\n",
      "\t\t<RecordStatus>current</RecordStatus>\n",
      "\t\t<Species>Homo sapiens</Species>\n",
      "\t\t<InterpretedRecord>\n",
      "\t\t\t<SimpleAllele AlleleID=\"206181\" VariationID=\"209223\">\n",
      "\t\t\t\t<GeneList>\n",
      "\t\t\t\t\t<Gene Symbol=\"BRCA1\" FullName=\"BRCA1 DNA repair associated\" GeneID=\"672\" HGNC_ID=\"HGNC:1100\" Source=\"submitted\" RelationshipType=\"near gene, downstream\">\n",
      "\t\t\t\t\t\t<Location>\n",
      "\t\t\t\t\t\t\t<CytogeneticLocation>17q21.31</CytogeneticLocation>\n",
      "\t\t\t\t\t\t\t<SequenceLocation Assembly=\"GRCh38\" AssemblyAccessionVersion=\"GCF_000001405.38\" AssemblyStatus=\"current\" Chr=\"17\" Accession=\"NC_000017.11\" start=\"43044295\" stop=\"43125364\" display_start=\"43044295\" display_stop=\"43125364\" Strand=\"-\" />\n",
      "\t\t\t\t\t\t\t<SequenceLocation Assembly=\"GRCh37\" AssemblyAccessionVersion=\"GCF_000001405.25\" AssemblyStatus=\"previous\" Chr=\"17\" Accession=\"NC_000017.10\" start=\"41196311\" stop=\"41277499\" display_start=\"41196311\" display_stop=\"41277499\" Strand=\"-\" />\n",
      "\t\t\t\t\t\t</Location>\n",
      "\t\t\t\t\t\t<OMIM>113705</OMIM>\n",
      "\t\t\t\t\t\t<Haploinsufficiency last_evaluated=\"2021-09-23\" ClinGen=\"https://www.ncbi.nlm.nih.gov/projects/dbvar/ISCA/isca_gene.cgi?sym=BRCA1\">Sufficient evidence for dosage pathogenicity</Haploinsufficiency>\n",
      "\t\t\t\t\t\t<Triplosensitivity last_evaluated=\"2021-09-23\" ClinGen=\"https://www.ncbi.nlm.nih.gov/projects/dbvar/ISCA/isca_gene.cgi?sym=BRCA1\">No evidence available</Triplosensitivity>\n",
      "\t\t\t\t\t\t<Property>gene_acmg_incidental_2013</Property>\n",
      "\t\t\t\t\t\t<Property>gene_acmg_incidental_2016</Property>\n",
      "\t\t\t\t\t\t<Property>gene_acmg_incidental_2021</Property>\n",
      "\t\t\t\t\t\t<Property>gene_acmg_incidental_2022</Property>\n",
      "\t\t\t\t\t</Gene>\n",
      "\t\t\t\t</GeneList>\n",
      "\t\t\t\t<Name>NM_007294.3(BRCA1):c.*4549G&gt;C</Name>\n",
      "\t\t\t\t<CanonicalSPDI>NC_000017.11:43041128:C:G</CanonicalSPDI>\n",
      "\t\t\t\t<VariantType>single nucleotide variant</VariantType>\n",
      "\t\t\t\t<Location>\n",
      "\t\t\t\t\t<CytogeneticLocation>17q21.31</CytogeneticLocation>\n",
      "\t\t\t\t\t<SequenceLocation Assembly=\"GRCh38\" AssemblyAccessionVersion=\"GCF_000001405.38\" forDisplay=\"true\" AssemblyStatus=\"current\" Chr=\"17\" Accession=\"NC_000017.11\" start=\"43041129\" stop=\"43041129\" display_start=\"43041129\" display_stop=\"43041129\" variantLength=\"1\" positionVCF=\"43041129\" referenceAlleleVCF=\"C\" alternateAlleleVCF=\"G\" />\n",
      "\t\t\t\t\t<SequenceLocation Assembly=\"GRCh37\" AssemblyAccessionVersion=\"GCF_000001405.25\" AssemblyStatus=\"previous\" Chr=\"17\" Accession=\"NC_000017.10\" start=\"41193146\" stop=\"41193146\" display_start=\"41193146\" display_stop=\"41193146\" variantLength=\"1\" positionVCF=\"41193146\" referenceAlleleVCF=\"C\" alternateAlleleVCF=\"G\" />\n",
      "\t\t\t\t</Location>\n",
      "\t\t\t\t<OtherNameList>\n",
      "\t\t\t\t\t<Name>10260 G&gt;C</Name>\n",
      "\t\t\t\t</OtherNameList>\n",
      "\t\t\t\t<HGVSlist>\n",
      "\t\t\t\t\t<HGVS Type=\"genomic\">\n",
      "\t\t\t\t\t\t<NucleotideExpression sequenceAccessionVersion=\"LRG_292\" sequenceAccession=\"LRG_292\">\n",
      "\t\t\t\t\t\t\t<Expression>LRG_292:g.176855G&gt;C</Expression>\n",
      "\t\t\t\t\t\t</NucleotideExpression>\n",
      "\t\t\t\t\t</HGVS>\n",
      "\t\t\t\t\t<HGVS Type=\"coding\">\n",
      "\t\t\t\t\t\t<NucleotideExpression sequenceAccessionVersion=\"LRG_292t1\" sequenceAccession=\"LRG_292t1\">\n",
      "\t\t\t\t\t\t\t<Expression>LRG_292t1:c.*4549G&gt;C</Expression>\n",
      "\t\t\t\t\t\t</NucleotideExpression>\n",
      "\t\t\t\t\t</HGVS>\n",
      "\t\t\t\t\t<HGVS Assembly=\"GRCh37\" Type=\"genomic, top-level\">\n",
      "\t\t\t\t\t\t<NucleotideExpression sequenceAccessionVersion=\"NC_000017.10\" sequenceAccession=\"NC_000017\" sequenceVersion=\"10\" change=\"g.41193146C&gt;G\" Assembly=\"GRCh37\">\n",
      "\t\t\t\t\t\t\t<Expression>NC_000017.10:g.41193146C&gt;G</Expression>\n",
      "\t\t\t\t\t\t</NucleotideExpression>\n",
      "\t\t\t\t\t</HGVS>\n",
      "\t\t\t\t\t<HGVS Assembly=\"GRCh38\" Type=\"genomic, top-level\">\n",
      "\t\t\t\t\t\t<NucleotideExpression sequenceAccessionVersion=\"NC_000017.11\" sequenceAccession=\"NC_000017\" sequenceVersion=\"11\" change=\"g.43041129C&gt;G\" Assembly=\"GRCh38\">\n",
      "\t\t\t\t\t\t\t<Expression>NC_000017.11:g.43041129C&gt;G</Expression>\n",
      "\t\t\t\t\t\t</NucleotideExpression>\n",
      "\t\t\t\t\t</HGVS>\n",
      "\t\t\t\t\t<HGVS Type=\"genomic\">\n",
      "\t\t\t\t\t\t<NucleotideExpression sequenceAccessionVersion=\"NG_005905.2\" sequenceAccession=\"NG_005905\" sequenceVersion=\"2\" change=\"g.176855G&gt;C\">\n",
      "\t\t\t\t\t\t\t<Expression>NG_005905.2:g.176855G&gt;C</Expression>\n",
      "\t\t\t\t\t\t</NucleotideExpression>\n",
      "\t\t\t\t\t</HGVS>\n",
      "\t\t\t\t</HGVSlist>\n",
      "\t\t\t\t<XRefList>\n",
      "\t\t\t\t\t<XRef ID=\"CA276199\" DB=\"ClinGen\" />\n",
      "\t\t\t\t\t<XRef Type=\"rs\" ID=\"138528005\" DB=\"dbSNP\" />\n",
      "\t\t\t\t</XRefList>\n",
      "\t\t\t\t<AlleleFrequencyList>\n",
      "\t\t\t\t\t<AlleleFrequency Value=\"0.00158\" Source=\"The Genome Aggregation Database (gnomAD)\" />\n",
      "\t\t\t\t\t<AlleleFrequency Value=\"0.00149\" Source=\"Trans-Omics for Precision Medicine (TOPMed)\" />\n",
      "\t\t\t\t\t<AlleleFrequency Value=\"0.00200\" Source=\"1000 Genomes Project\" />\n",
      "\t\t\t\t</AlleleFrequencyList>\n",
      "\t\t\t\t<GlobalMinorAlleleFrequency Value=\"0.00200\" Source=\"1000 Genomes Project\" MinorAllele=\"G\" />\n",
      "\t\t\t</SimpleAllele>\n",
      "\t\t\t<ReviewStatus>reviewed by expert panel</ReviewStatus>\n",
      "\t\t\t<RCVList>\n",
      "\t\t\t\t<RCVAccession Title=\"NM_007294.3(BRCA1):c.*4549G&gt;C AND Breast-ovarian cancer, familial, susceptibility to, 1\" DateLastEvaluated=\"2015-01-12\" ReviewStatus=\"reviewed by expert panel\" Interpretation=\"Benign\" SubmissionCount=\"1\" Accession=\"RCV000191166\" Version=\"3\">\n",
      "\t\t\t\t\t<InterpretedConditionList TraitSetID=\"1917\">\n",
      "\t\t\t\t\t\t<InterpretedCondition DB=\"MedGen\" ID=\"C2676676\">Breast-ovarian cancer, familial, susceptibility to, 1</InterpretedCondition>\n",
      "\t\t\t\t\t</InterpretedConditionList>\n",
      "\t\t\t\t</RCVAccession>\n",
      "\t\t\t</RCVList>\n",
      "\t\t\t<Interpretations>\n",
      "\t\t\t\t<Interpretation DateLastEvaluated=\"2015-01-12\" NumberOfSubmissions=\"1\" NumberOfSubmitters=\"1\" Type=\"Clinical significance\">\n",
      "\t\t\t\t\t<Description>Benign</Description>\n",
      "\t\t\t\t\t<ConditionList>\n",
      "\t\t\t\t\t\t<TraitSet ID=\"1917\" Type=\"Disease\" ContributesToAggregateClinsig=\"true\">\n",
      "\t\t\t\t\t\t\t<Trait ID=\"4711\" Type=\"Disease\">\n",
      "\t\t\t\t\t\t\t\t<Name>\n",
      "\t\t\t\t\t\t\t\t\t<ElementValue Type=\"Alternate\">Breast-ovarian cancer, familial 1</ElementValue>\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"Breast-ovarian+cancer%2C+familial+1/7865\" DB=\"Genetic Alliance\" />\n",
      "\t\t\t\t\t\t\t\t</Name>\n",
      "\t\t\t\t\t\t\t\t<Name>\n",
      "\t\t\t\t\t\t\t\t\t<ElementValue Type=\"Alternate\">BREAST CANCER, FAMILIAL, SUSCEPTIBILITY TO, 1</ElementValue>\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"604370\" DB=\"OMIM\" />\n",
      "\t\t\t\t\t\t\t\t</Name>\n",
      "\t\t\t\t\t\t\t\t<Name>\n",
      "\t\t\t\t\t\t\t\t\t<ElementValue Type=\"Alternate\">Breast cancer, familial 1</ElementValue>\n",
      "\t\t\t\t\t\t\t\t</Name>\n",
      "\t\t\t\t\t\t\t\t<Name>\n",
      "\t\t\t\t\t\t\t\t\t<ElementValue Type=\"Alternate\">OVARIAN CANCER, SUSCEPTIBILITY TO</ElementValue>\n",
      "\t\t\t\t\t\t\t\t\t<XRef Type=\"Allelic variant\" ID=\"602667.0001\" DB=\"OMIM\" />\n",
      "\t\t\t\t\t\t\t\t</Name>\n",
      "\t\t\t\t\t\t\t\t<Name>\n",
      "\t\t\t\t\t\t\t\t\t<ElementValue Type=\"Preferred\">Breast-ovarian cancer, familial, susceptibility to, 1</ElementValue>\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"MONDO:0011450\" DB=\"MONDO\" />\n",
      "\t\t\t\t\t\t\t\t</Name>\n",
      "\t\t\t\t\t\t\t\t<Symbol>\n",
      "\t\t\t\t\t\t\t\t\t<ElementValue Type=\"Alternate\">HBOC1</ElementValue>\n",
      "\t\t\t\t\t\t\t\t\t<XRef Type=\"MIM\" ID=\"604370\" DB=\"OMIM\" />\n",
      "\t\t\t\t\t\t\t\t</Symbol>\n",
      "\t\t\t\t\t\t\t\t<Symbol>\n",
      "\t\t\t\t\t\t\t\t\t<ElementValue Type=\"Alternate\">BROVCA1</ElementValue>\n",
      "\t\t\t\t\t\t\t\t\t<XRef Type=\"MIM\" ID=\"604370\" DB=\"OMIM\" />\n",
      "\t\t\t\t\t\t\t\t</Symbol>\n",
      "\t\t\t\t\t\t\t\t<Symbol>\n",
      "\t\t\t\t\t\t\t\t\t<ElementValue Type=\"Alternate\">HBOC</ElementValue>\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000501743\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t</Symbol>\n",
      "\t\t\t\t\t\t\t\t<AttributeSet>\n",
      "\t\t\t\t\t\t\t\t\t<Attribute Type=\"public definition\">BRCA1- and BRCA2-associated hereditary breast and ovarian cancer (HBOC) is characterized by an increased risk for female and male breast cancer, ovarian cancer (including fallopian tube and primary peritoneal cancers), and to a lesser extent other cancers such as prostate cancer, pancreatic cancer, and melanoma primarily in individuals with a BRCA2 pathogenic variant. The risk of developing an associated cancer varies depending on whether HBOC is caused by a BRCA1 or BRCA2 pathogenic variant.</Attribute>\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"NBK1247\" DB=\"GeneReviews\" />\n",
      "\t\t\t\t\t\t\t\t</AttributeSet>\n",
      "\t\t\t\t\t\t\t\t<AttributeSet>\n",
      "\t\t\t\t\t\t\t\t\t<Attribute Type=\"keyword\">Neoplasm</Attribute>\n",
      "\t\t\t\t\t\t\t\t</AttributeSet>\n",
      "\t\t\t\t\t\t\t\t<AttributeSet>\n",
      "\t\t\t\t\t\t\t\t\t<Attribute Type=\"disease mechanism\" integerValue=\"273\">loss of function</Attribute>\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000501746\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000507864\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000509450\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000509692\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000509982\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000512644\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000512645\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000520872\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000552245\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000569406\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000021517\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000320777\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000332529\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000509001\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000509983\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000512320\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000520865\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000520871\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000522160\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000527942\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000530120\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000552304\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000562228\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000569423\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000570011\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000325409\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000501817\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000505644\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000507653\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000512816\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000514601\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000519030\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000520069\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000520869\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000522161\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000528915\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000531275\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000551440\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000017876\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000330054\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000501743\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000507764\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000509002\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000509980\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000520071\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000520072\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000520394\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000520866\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000520867\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000520870\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000521908\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000522159\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000530202\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t\t<XRef ID=\"GTR000592961\" DB=\"Genetic Testing Registry (GTR)\" />\n",
      "\t\t\t\t\t\t\t\t</AttributeSet>\n",
      "\t\t\t\t\t\t\t\t<Citation Type=\"review\" Abbrev=\"GeneReviews\">\n",
      "\t\t\t\t\t\t\t\t\t<ID Source=\"PubMed\">20301425</ID>\n",
      "\t\t\t\t\t\t\t\t\t<ID Source=\"BookShelf\">NBK1247</ID>\n",
      "\t\t\t\t\t\t\t\t</Citation>\n",
      "\t\t\t\t\t\t\t\t<Citation Type=\"practice guideline\" Abbrev=\"ACS, 2007\">\n",
      "\t\t\t\t\t\t\t\t\t<ID Source=\"PubMed\">17392385</ID>\n",
      "\t\t\t\t\t\t\t\t</Citation>\n",
      "\t\t\t\t\t\t\t\t<Citation Type=\"Position Statement\" Abbrev=\"ASCO, 2010\">\n",
      "\t\t\t\t\t\t\t\t\t<ID Source=\"PubMed\">20065170</ID>\n",
      "\t\t\t\t\t\t\t\t</Citation>\n",
      "\t\t\t\t\t\t\t\t<Citation Type=\"practice guideline\" Abbrev=\"NSGC, 2004\">\n",
      "\t\t\t\t\t\t\t\t\t<ID Source=\"PubMed\">15604628</ID>\n",
      "\t\t\t\t\t\t\t\t</Citation>\n",
      "\t\t\t\t\t\t\t\t<Citation Type=\"practice guideline\" Abbrev=\"NSGC, 2007\">\n",
      "\t\t\t\t\t\t\t\t\t<ID Source=\"PubMed\">17508274</ID>\n",
      "\t\t\t\t\t\t\t\t</Citation>\n",
      "\t\t\t\t\t\t\t\t<Citation Type=\"practice guideline\" Abbrev=\"ACOG, 2009\">\n",
      "\t\t\t\t\t\t\t\t\t<ID Source=\"PubMed\">19305347</ID>\n",
      "\t\t\t\t\t\t\t\t</Citation>\n",
      "\t\t\t\t\t\t\t\t<Citation Type=\"Recommendation\" Abbrev=\"ACMG, 2013\">\n",
      "\t\t\t\t\t\t\t\t\t<ID Source=\"PubMed\">23788249</ID>\n",
      "\t\t\t\t\t\t\t\t</Citation>\n",
      "\t\t\t\t\t\t\t\t<Citation Type=\"Suggested Reading\" Abbrev=\"Phillips et al., 2013\">\n",
      "\t\t\t\t\t\t\t\t\t<ID Source=\"PubMed\">23918944</ID>\n",
      "\t\t\t\t\t\t\t\t</Citation>\n",
      "\t\t\t\t\t\t\t\t<Citation Type=\"Suggested Reading\" Abbrev=\"Domchek et al., 2010\">\n",
      "\t\t\t\t\t\t\t\t\t<ID Source=\"pmc\">2948529</ID>\n",
      "\t\t\t\t\t\t\t\t</Citation>\n",
      "\t\t\t\t\t\t\t\t<Citation Type=\"practice guideline\" Abbrev=\"ASCO, 2014\">\n",
      "\t\t\t\t\t\t\t\t\t<ID Source=\"PubMed\">24493721</ID>\n",
      "\t\t\t\t\t\t\t\t</Citation>\n",
      "\t\t\t\t\t\t\t\t<Citation Type=\"practice guideline\" Abbrev=\"USPSTF, 2014 [Archived]\">\n",
      "\t\t\t\t\t\t\t\t\t<ID Source=\"PubMed\">24366376</ID>\n",
      "\t\t\t\t\t\t\t\t</Citation>\n",
      "\t\t\t\t\t\t\t\t<Citation Type=\"practice guideline\" Abbrev=\"NSGC, 2013\">\n",
      "\t\t\t\t\t\t\t\t\t<ID Source=\"PubMed\">23188549</ID>\n",
      "\t\t\t\t\t\t\t\t</Citation>\n",
      "\t\t\t\t\t\t\t\t<Citation Type=\"practice guideline\" Abbrev=\"ACMG/NSGC, 2015\">\n",
      "\t\t\t\t\t\t\t\t\t<ID Source=\"PubMed\">25394175</ID>\n",
      "\t\t\t\t\t\t\t\t</Citation>\n",
      "\t\t\t\t\t\t\t\t<Citation Type=\"Recommendation\" Abbrev=\"USPSTF, 2014\">\n",
      "\t\t\t\t\t\t\t\t\t<ID Source=\"PubMed\">24366402</ID>\n",
      "\t\t\t\t\t\t\t\t</Citation>\n",
      "\t\t\t\t\t\t\t\t<Citation Type=\"Suggested Reading\" Abbrev=\"USPSTF, 2013\">\n",
      "\t\t\t\t\t\t\t\t\t<ID Source=\"PubMed\">24432435</ID>\n",
      "\t\t\t\t\t\t\t\t</Citation>\n",
      "\t\t\t\t\t\t\t\t<Citation Type=\"Recommendation\" Abbrev=\"ACMG, 2015\">\n",
      "\t\t\t\t\t\t\t\t\t<ID Source=\"PubMed\">25356965</ID>\n",
      "\t\t\t\t\t\t\t\t</Citation>\n",
      "\t\t\t\t\t\t\t\t<Citation Type=\"Recommendation\" Abbrev=\"ACMG, 2016\">\n",
      "\t\t\t\t\t\t\t\t\t<ID Source=\"PubMed\">27854360</ID>\n",
      "\t\t\t\t\t\t\t\t\t<ID Source=\"DOI\">10.1038/gim.2016.190</ID>\n",
      "\t\t\t\t\t\t\t\t</Citation>\n",
      "\t\t\t\t\t\t\t\t<Citation Type=\"Suggested Reading\" Abbrev=\"NCI PDQ, Cancer Genetics Counseling\">\n",
      "\t\t\t\t\t\t\t\t\t<ID Source=\"PubMed\">26389258</ID>\n",
      "\t\t\t\t\t\t\t\t</Citation>\n",
      "\t\t\t\t\t\t\t\t<Citation Type=\"Recommendation\" Abbrev=\"ACMG SF v3.0, 2021\">\n",
      "\t\t\t\t\t\t\t\t\t<ID Source=\"PubMed\">34012068</ID>\n",
      "\t\t\t\t\t\t\t\t\t<ID Source=\"DOI\">10.1038/s41436-021-01172-3</ID>\n",
      "\t\t\t\t\t\t\t\t</Citation>\n",
      "\t\t\t\t\t\t\t\t<Citation Type=\"practice guideline\" Abbrev=\"NSGC, 2021\">\n",
      "\t\t\t\t\t\t\t\t\t<ID Source=\"PubMed\">33410258</ID>\n",
      "\t\t\t\t\t\t\t\t</Citation>\n",
      "\t\t\t\t\t\t\t\t<Citation Type=\"Recommendation\" Abbrev=\"ACMG SF v3.1, 2022\">\n",
      "\t\t\t\t\t\t\t\t\t<ID Source=\"PubMed\">35802134</ID>\n",
      "\t\t\t\t\t\t\t\t\t<ID Source=\"DOI\">10.1016/j.gim.2022.04.006</ID>\n",
      "\t\t\t\t\t\t\t\t</Citation>\n",
      "\t\t\t\t\t\t\t\t<Citation Type=\"Position Statement\" Abbrev=\"ASCO, 2015\">\n",
      "\t\t\t\t\t\t\t\t\t<ID Source=\"PubMed\">26324357</ID>\n",
      "\t\t\t\t\t\t\t\t</Citation>\n",
      "\t\t\t\t\t\t\t\t<Citation Type=\"practice guideline\" Abbrev=\"NCCN, 2022\">\n",
      "\t\t\t\t\t\t\t\t\t<URL>https://www.nccn.org/professionals/physician_gls/pdf/genetics_bop.pdf</URL>\n",
      "\t\t\t\t\t\t\t\t\t<CitationText>NCCN Clinical Practice Guidelines in Oncology, Genetic/Familial High-Risk Assessment: Breast, Ovarian, and Pancreatic version 2.2022</CitationText>\n",
      "\t\t\t\t\t\t\t\t</Citation>\n",
      "\t\t\t\t\t\t\t\t<Citation Type=\"Recommendation\" Abbrev=\"USPSTF, 2019\">\n",
      "\t\t\t\t\t\t\t\t\t<ID Source=\"PubMed\">31429903</ID>\n",
      "\t\t\t\t\t\t\t\t</Citation>\n",
      "\t\t\t\t\t\t\t\t<Citation Type=\"Suggested Reading\" Abbrev=\"USPSTF, 2019\">\n",
      "\t\t\t\t\t\t\t\t\t<ID Source=\"PubMed\">31479213</ID>\n",
      "\t\t\t\t\t\t\t\t\t<ID Source=\"BookShelf\">NBK545867</ID>\n",
      "\t\t\t\t\t\t\t\t</Citation>\n",
      "\t\t\t\t\t\t\t\t<XRef ID=\"145\" DB=\"Orphanet\" />\n",
      "\t\t\t\t\t\t\t\t<XRef ID=\"C2676676\" DB=\"MedGen\" />\n",
      "\t\t\t\t\t\t\t\t<XRef ID=\"MONDO:0011450\" DB=\"MONDO\" />\n",
      "\t\t\t\t\t\t\t\t<XRef Type=\"MIM\" ID=\"604370\" DB=\"OMIM\" />\n",
      "\t\t\t\t\t\t\t</Trait>\n",
      "\t\t\t\t\t\t</TraitSet>\n",
      "\t\t\t\t\t</ConditionList>\n",
      "\t\t\t\t</Interpretation>\n",
      "\t\t\t</Interpretations>\n",
      "\t\t\t<ClinicalAssertionList>\n",
      "\t\t\t\t<ClinicalAssertion ID=\"456000\" SubmissionDate=\"2015-08-17\" DateLastUpdated=\"2015-09-29\" DateCreated=\"2015-09-29\">\n",
      "\t\t\t\t\t<ClinVarSubmissionID localKey=\"NM_007294.3:c.*4549G&gt;C|OMIM:604370\" />\n",
      "\t\t\t\t\t<ClinVarAccession Accession=\"SCV000244504\" DateUpdated=\"2015-09-29\" DateCreated=\"2015-09-29\" Type=\"SCV\" Version=\"1\" SubmitterName=\"Evidence-based Network for the Interpretation of Germline Mutant Alleles (ENIGMA)\" OrgID=\"504863\" OrganizationCategory=\"consortium\" OrgAbbreviation=\"ENIGMA\" />\n",
      "\t\t\t\t\t<RecordStatus>current</RecordStatus>\n",
      "\t\t\t\t\t<ReviewStatus>reviewed by expert panel</ReviewStatus>\n",
      "\t\t\t\t\t<Interpretation DateLastEvaluated=\"2015-01-12\">\n",
      "\t\t\t\t\t\t<Description>Benign</Description>\n",
      "\t\t\t\t\t\t<Comment>Class 1 not pathogenic based on frequency &gt;1% in an outbred sampleset. Frequency 0.02 (African), derived from 1000 genomes (2012-04-30).</Comment>\n",
      "\t\t\t\t\t</Interpretation>\n",
      "\t\t\t\t\t<Assertion>variation to disease</Assertion>\n",
      "\t\t\t\t\t<AttributeSet>\n",
      "\t\t\t\t\t\t<Attribute Type=\"AssertionMethod\">ENIGMA BRCA1/2 Classification Criteria (2015)</Attribute>\n",
      "\t\t\t\t\t\t<Citation Type=\"general\">\n",
      "\t\t\t\t\t\t\t<URL>https://submit.ncbi.nlm.nih.gov/ft/byid/c2ffjci6/enigma_rules_2015-03-26.pdf</URL>\n",
      "\t\t\t\t\t\t\t<CitationText>ENIGMA BRCA1/2 Classification Criteria (2015)</CitationText>\n",
      "\t\t\t\t\t\t</Citation>\n",
      "\t\t\t\t\t</AttributeSet>\n",
      "\t\t\t\t\t<ObservedInList>\n",
      "\t\t\t\t\t\t<ObservedIn>\n",
      "\t\t\t\t\t\t\t<Sample>\n",
      "\t\t\t\t\t\t\t\t<Origin>germline</Origin>\n",
      "\t\t\t\t\t\t\t\t<Species TaxonomyId=\"9606\">human</Species>\n",
      "\t\t\t\t\t\t\t\t<AffectedStatus>unknown</AffectedStatus>\n",
      "\t\t\t\t\t\t\t</Sample>\n",
      "\t\t\t\t\t\t\t<Method>\n",
      "\t\t\t\t\t\t\t\t<MethodType>curation</MethodType>\n",
      "\t\t\t\t\t\t\t</Method>\n",
      "\t\t\t\t\t\t\t<ObservedData>\n",
      "\t\t\t\t\t\t\t\t<Attribute Type=\"Description\">not provided</Attribute>\n",
      "\t\t\t\t\t\t\t</ObservedData>\n",
      "\t\t\t\t\t\t</ObservedIn>\n",
      "\t\t\t\t\t</ObservedInList>\n",
      "\t\t\t\t\t<SimpleAllele>\n",
      "\t\t\t\t\t\t<GeneList>\n",
      "\t\t\t\t\t\t\t<Gene Symbol=\"BRCA1\" />\n",
      "\t\t\t\t\t\t</GeneList>\n",
      "\t\t\t\t\t\t<VariantType>Variation</VariantType>\n",
      "\t\t\t\t\t\t<OtherNameList>\n",
      "\t\t\t\t\t\t\t<Name>10260 G&gt;C</Name>\n",
      "\t\t\t\t\t\t</OtherNameList>\n",
      "\t\t\t\t\t\t<AttributeSet>\n",
      "\t\t\t\t\t\t\t<Attribute Type=\"HGVS\">NM_007294.3:c.*4549G&gt;C</Attribute>\n",
      "\t\t\t\t\t\t</AttributeSet>\n",
      "\t\t\t\t\t</SimpleAllele>\n",
      "\t\t\t\t\t<TraitSet Type=\"Disease\">\n",
      "\t\t\t\t\t\t<Trait Type=\"Disease\">\n",
      "\t\t\t\t\t\t\t<Name>\n",
      "\t\t\t\t\t\t\t\t<ElementValue Type=\"Preferred\">Breast-ovarian cancer, familial 1</ElementValue>\n",
      "\t\t\t\t\t\t\t</Name>\n",
      "\t\t\t\t\t\t\t<XRef DB=\"OMIM\" ID=\"604370\" Type=\"MIM\" />\n",
      "\t\t\t\t\t\t</Trait>\n",
      "\t\t\t\t\t</TraitSet>\n",
      "\t\t\t\t\t<StudyDescription>ENIGMA (Evidence-based Network for the Interpretation of Germline Mutant Alleles) is a consortium focused on determining the clinical significance of variants in BRCA1, BRCA2 and other known/suspected breast cancer genes. http://enigmaconsortium.org/</StudyDescription>\n",
      "\t\t\t\t\t<SubmissionNameList>\n",
      "\t\t\t\t\t\t<SubmissionName>ENIGMA_2015-08-17</SubmissionName>\n",
      "\t\t\t\t\t</SubmissionNameList>\n",
      "\t\t\t\t</ClinicalAssertion>\n",
      "\t\t\t</ClinicalAssertionList>\n",
      "\t\t\t<TraitMappingList>\n",
      "\t\t\t\t<TraitMapping ClinicalAssertionID=\"456000\" TraitType=\"Disease\" MappingType=\"Name\" MappingValue=\"Breast-ovarian cancer, familial 1\" MappingRef=\"Preferred\">\n",
      "\t\t\t\t\t<MedGen CUI=\"C2676676\" Name=\"Breast-ovarian cancer, familial, susceptibility to, 1\" />\n",
      "\t\t\t\t</TraitMapping>\n",
      "\t\t\t</TraitMappingList>\n",
      "\t\t</InterpretedRecord>\n",
      "\t</VariationArchive>\n",
      "</ClinVarResult-Set>\n"
     ]
    }
   ],
   "source": [
    "def _pretty_print(current, parent=None, index=-1, depth=0):\n",
    "    for i, node in enumerate(current):\n",
    "        _pretty_print(node, current, i, depth + 1)\n",
    "    if parent is not None:\n",
    "        if index == 0:\n",
    "            parent.text = '\\n' + ('\\t' * depth)\n",
    "        else:\n",
    "            parent[index - 1].tail = '\\n' + ('\\t' * depth)\n",
    "        if index == len(parent) - 1:\n",
    "            current.tail = '\\n' + ('\\t' * (depth - 1))\n",
    "\n",
    "_pretty_print(root)\n",
    "\n",
    "ET.dump(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "root = result.getroot()\n",
    "root.attrib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<Element 'SequenceLocation' at 0x0000025029C244F0>\n",
      "SequenceLocation\n",
      "{'Assembly': 'GRCh38', 'AssemblyAccessionVersion': 'GCF_000001405.38', 'forDisplay': 'true', 'AssemblyStatus': 'current', 'Chr': '17', 'Accession': 'NC_000017.11', 'start': '43041129', 'stop': '43041129', 'display_start': '43041129', 'display_stop': '43041129', 'variantLength': '1', 'positionVCF': '43041129', 'referenceAlleleVCF': 'C', 'alternateAlleleVCF': 'G'}\n",
      "GRCh38\n"
     ]
    }
   ],
   "source": [
    "var_loc = root.findall('./VariationArchive/InterpretedRecord/SimpleAllele/Location/SequenceLocation')\n",
    "\n",
    "for info in var_loc:\n",
    "    print(info)\n",
    "    print(info.tag)\n",
    "    print(info.attrib)\n",
    "    print(info.attrib['Assembly'])\n",
    "    if info.attrib['Assembly'] == 'GRCh38':\n",
    "        chr_acc = info.attrib['Accession']\n",
    "        start   = int(info.attrib['start'])\n",
    "        ref_nt  = info.attrib['referenceAlleleVCF']\n",
    "        alt_nt  = info.attrib['alternateAlleleVCF']\n",
    "\n",
    "        break\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TCTGCTCCTAGATTGCTAATATAATTTAATCTTTGTCCTTTGCATTGTACAGCAGTCCCCCCTTATCTGCGGGGCATATGTTCCAAGACCACAAGTAGATGCCTGAACCCTCGGATAGTAC\n"
     ]
    }
   ],
   "source": [
    "chr_seq_fetch = Entrez.efetch(db=\"nucleotide\", \n",
    "                              id=\"NC_000017\", \n",
    "                              rettype=\"fasta\", \n",
    "                              strand=1, \n",
    "                              seq_start=start-60, \n",
    "                              seq_stop=start+60)\n",
    "\n",
    "record = SeqIO.read(chr_seq_fetch, \"fasta\")\n",
    "chr_seq_fetch.close()\n",
    "print(record.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.13 ('genet')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "ad333f69f1d349cc4ee014c46a9ac73e8a7331a44020b2a0d8d29b366c987199"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
